<!DOCTYPE html>
<html lang="">

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/pure/0.6.0/pure-min.css">

    <!--[if lte IE 8]>
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/pure/0.6.0/grids-responsive-old-ie-min.css">
    <![endif]-->
    <!--[if gt IE 8]><!-->
        <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/pure/0.6.0/grids-responsive-min.css">
    <!--<![endif]-->

    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.3.0/styles/default.min.css">
    <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.3.0/highlight.min.js"></script>
    <script   src="https://code.jquery.com/jquery-3.1.0.min.js"   integrity="sha256-cCueBR6CsyA4/9szpPfrX3s49M9vUU5BgtiJj06wt/s="   crossorigin="anonymous"></script>
    <script>hljs.initHighlightingOnLoad();</script>

    <title>HiGlass: Multi-resolution Hi-C data display</title>
</head>

<script src="https://rawgit.com/pkerpedjiev/goomba/master/dist/scripts/goomba.js"></script>

<!-- build:css styles/style.css -->
<link rel="stylesheet" href="styles/page.css" />
<!-- endbuild -->

<!-- build:js scripts/higlass.js -->
<script src='scripts/higlass.js'></script>
<!-- endbuild -->

<body>
    <div style='margin-left: 20px;'>
            <h2>Zoomable gene tracks</h2>

            We can combine gene annotations, wiggle tracks and the heatmap into one
            integrated display:
            <br>
            <br>
            <div class="ui-widget" id='dataset-origin-div' style="margin-left: 30px;">
                <table>
                    <tr>
                        <td>
                Data Source: 
                        </td>
                        <td>
                            <input type="text"  id="dataset-origin-text" value="//52.23.165.123:9872/hg19/Rao2014-GM12878-MboI-allreps-filtered.1kb.cool.reduced.genome.gz" onkeypress="keyPressed(event)" style="left:30px; width: 380px">
                            <span class="input-group-btn">
                                <button class="btn btn-default" type="button" id='position-search-button' onclick="loadDataset()">Go!</button>
                            </span>
                        </td>
                    </tr>
                    <tr>
                        <td>
                            Search Location:
                        </td>
                        <td>
                            <input type="text"  id="position-search-text" value="chr1 to chr1:100000 and chr2:10 to chr2:10000" onkeypress="keyPressed(event)" style="left:30px">
                            <span class="input-group-btn">
                                <button class="btn btn-default" type="button" id='position-search-button' onclick="searchPosition()">Go!</button>
                            </span>
                        </td>
                    </tr>
                </table>
            </div>
            <br>
    <svg id="goombaPlot" style="position:relative; height:100px; width: 700px" >
    <div id="mmv-area" style="position:relative; height:550px; width: 700px" >
    </div>
    </div>
</body>


<script type='text/javascript'>

/*
var awsDomain = '//52.23.165.123:9872';
var assembly = 'mm9'
var higlassDescription = {
    trackHeight: 25,
    chromInfoFile: 'https://s3.amazonaws.com/pkerp/data/' + assembly + '/chromInfo.txt',
    tileTracks: [{ 'type': 'GeneTrack',
                   'data': awsDomain + '/' + assembly + '/refgene-tiles-minus'},
                 { 'type': 'GeneTrack',
                   'data': awsDomain + '/' + assembly + '/refgene-tiles-plus'}]
}
*/
var searchField = null;

function createHiglass(awsDomain, assembly, datasetName) {
        var width = 700, height=500;


        d3.select('#mmv-area').selectAll('*').remove();
        var svg = d3.select('#goombaPlot')
        .attr('width', width)
        .attr('height', 300);

        svg.selectAll('*').remove();

        /*
        // goomba gets drawn with a margin of 30 on the left
        // and a margin of 30 on the right
        // whereas higlass gets drawn with a margin of 30 on the left
        // and 120 on the right
        // thus goomba needs to be 90 smaller
            var goombaPlot = goomba.Goomba()
                .width(width - 90)
                .height(80)

                var gMain = svg.append('g')
                .classed('g-main', true)
                .datum('//s3.amazonaws.com/pkerp/data/hg19/refgene-tiles-small')
                //.datum('jsons/tiles')
                .call(goombaPlot)
        */


    var trackHeight = 25;
    //var awsDomain = '//search-higlass-ssxwuix6kow3sekyeresi7ay5e.us-east-1.es.amazonaws.com';
    var tileDirectories = ['//' + awsDomain + '/' + assembly + '/refgene-tiles-minus',
                           '//' + awsDomain + '/' + assembly + '/refgene-tiles-plus']
    var chromInfoFile = 'https://s3.amazonaws.com/pkerp/data/' + assembly + '/chromInfo.txt';
                           


    var margin = {'top': 20, 'left': 30, 'bottom': 30, 'right': 120};
    var mmvMargin = {'top': 10, 'left': 30, 'bottom': 30, 'right': 120};

    goomba.ChromosomeInfo(chromInfoFile, function(chromInfo) {
    ///////////////////////////
        //var xScale = d3.scale.linear().domain([0, 5000000])
        console.log('chromInfo:', chromInfo);
        var xScale = d3.scale.linear().domain([0, chromInfo.totalLength])
            .range([0, width - mmvMargin.left - mmvMargin.right]);

        var zoomDispatch = d3.dispatch('zoom', 'zoomend');
        searchField = new goomba.SearchField(chromInfo, xScale.copy(),
                xScale.copy(), zoomDispatch);

        var mmvPlot1 = higlass.MassiveMatrixPlot()
            .width(width)
            .height(height)
            .nTicks(4)
            .margin(mmvMargin)
            .xDomain(xScale.domain())
            .yDomain(xScale.domain())
            .zoomDispatch(zoomDispatch)
            .chromInfo(chromInfo);

            d3.select('#mmv-area')
            .datum( '//' + awsDomain + '/' + assembly + '/' + datasetName )
            //.datum( awsDomain + '/hg19/Rao2014-GM12878-MboI-allreps-filtered.1kb.cool.reduced.genome.mmmili.gz')
            //.datum( awsDomain + '/hg19/Dixon2015-H1hESC_ES-HindIII-allreps-filtered.1kb.cool.reduced.genome.gz')
            .call(mmvPlot1)

        var chromInfoPlot = goomba.ChromosomeAxis()
        .width(width - margin.left - margin.right)
        .domain(xScale.domain())
        .orient('top')
        .zoomDispatch(zoomDispatch);

        svg.append('g')
        .attr('transform', 'translate(' + margin.left + ',20)')
        .datum(chromInfo)
        .call(chromInfoPlot);

        var tileLayout = goomba.GeneTileLayout;

        var gMain = svg.append('g')
        .classed('g-main', true);

        /////////////////////////////
        var tiledArea = goomba.TiledArea().width(width)
        .height(height)
        .dataPointLayout(goomba.GenePlot)
        .tileLayout(tileLayout)
        .domain(xScale.domain())
        //.scaleExtent(scaleExtent)
        .margin({'left': 30, 'top': 50, 'right': 120, 'bottom': 10})
        .zoomDispatch(zoomDispatch);

        var gGeneTracks = gMain
            .append('g')
        .selectAll('g')
        .data(tileDirectories)
        .enter()
        .append('g')
        .attr('transform', function(d,i) { return 'translate(0,' + i * trackHeight + ')'; })
        .classed('track-g', true)
        .call(tiledArea)

        var zoomableLabels = goomba.ZoomableLabels()
        .markerClass('.gene-marker')
        .labelClass('.gene-label')
        .labelParent(gMain)
        .labelMarkerId(function(d) { return `n-${d.uid}`})
        .uidString('uid')

        tiledArea.on('draw', function() {
            gGeneTracks.call(zoomableLabels);
        });
        //////////////////////////////////////////

        var wiggleTileArea = goomba.TiledArea().width(width)
        .height(height)
        .dataPointLayout(goomba.GenePlot)
        .tileLayout(goomba.WiggleTileLayout)
        .domain(xScale.domain())
        .margin({'left': 30, 'top': 65, 'right': 120, 'bottom': 10})
        .zoomDispatch(zoomDispatch);

        var gWiggleTracks = gMain
        .append('g')
        .attr('transform', 'translate(0,' + (20) + ')')
        .selectAll('g')
        .data([
        /*
        ['https://s3.amazonaws.com/pkerp/data/served/hg19/E116-DNase.fc.signal.bigwig'])
        .data([awsDomain + '/hg19/wgEncodeSydhTfbsGm12878Ctcfsc15914c20StdSig.bigWig.bedGraph.genome.sorted.gz',
        awsDomain + '/hg19/E116-DNase.fc.signal.bigwig.bedGraph.genome.sorted.gz'
        */
        ])

        .enter()
        .append('g')
        .attr('transform', function(d,i) { return 'translate(0,' + (i * trackHeight + margin.top) + ')'; })
        .classed('track-g', true)
        .call(wiggleTileArea)
    });

    }

        function searchPosition() {
            var text = d3.select('#position-search-text').property('value');

            var pos = searchField.searchPosition(text);

            console.log('text:', text, 'pos:', pos);
        }

        function loadDataset() {
            var text = d3.select('#dataset-origin-text').property('value');
            console.log('text:', text);

            var parts = text.split('//')[1].split('/');
            var awsDomain = parts[0];
            var assembly = parts[parts.length-2];
            var datasetName = parts[parts.length-1];

            createHiglass(awsDomain, assembly, datasetName);
        }

/*
    function keyPressed(e) {
        if(e.keyCode === 13){
            console.log('enter pressed');
            searchPosition();
        }
    }
    */

createHiglass('52.23.165.123:9872', 'mm9', 'Dixon2012-J1-NcoI-R1-filtered.5kb.cool.reduced.genome.gz' );
///////////////////////////////////////////////


</script>
</body>
</html>
